src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,516 @@
+			, this, weak_from_this(), _1));
+}
+
+void websocket_tracker_connection::close()
+{
+	if (m_websocket)
+	{
+		m_websocket->close();
+		m_websocket.reset();
+	}
+
+	error_code const ec = error::operation_aborted;
+	while (!m_pending.empty())
+	{
+		auto [msg, callback] = std::move(m_pending.front());
+		TORRENT_UNUSED(msg);
+		m_pending.pop();
+		if (auto cb = callback.lock())
+			cb->tracker_request_error(
+					tracker_req()
+					, ec
+					, operation_t::unknown
+					, ec.message()
+					, seconds32(120));
+	}
+
+	m_callbacks.clear();
+	m_man.remove_request(this);
+}
+
+bool websocket_tracker_connection::is_started() const
+{
+	return m_websocket && (m_websocket->is_open() || m_websocket->is_connecting());
+}
+
+bool websocket_tracker_connection::is_open() const
+{
+	return m_websocket && m_websocket->is_open();
+}
+
+void websocket_tracker_connection::queue_request(tracker_request req, std::weak_ptr<request_callback> cb)
+{
+	m_pending.emplace(tracker_message{std::move(req)}, cb);
+	if (is_open()) send_pending();
+}
+
+void websocket_tracker_connection::queue_answer(tracker_answer ans)
+{
+	m_pending.emplace(tracker_message{std::move(ans)}, std::weak_ptr<request_callback>{});
+	if (is_open()) send_pending();
+}
+
+void websocket_tracker_connection::send_pending()
+{
+	if (m_sending || m_pending.empty()) return;
+
+	m_sending = true;
+
+	auto [msg, callback] = std::move(m_pending.front());
+	m_pending.pop();
+
+	std::visit([&](auto const& m)
+		{
+			// Update requester and store callback
+			if (callback.lock())
+			{
+				m_requester = callback;
+				m_callbacks[m.info_hash] = callback;
+			}
+
+			do_send(m);
+		}
+		, msg);
+}
+
+void websocket_tracker_connection::do_send(tracker_request const& req)
+{
+	// Update request
+	m_req = req;
+
+	json::object payload;
+	payload["action"] = "announce";
+	payload["info_hash"] = latin1_utf8(req.info_hash);
+	payload["uploaded"] = req.uploaded;
+	payload["downloaded"] = req.downloaded;
+	payload["left"] = req.left;
+	payload["corrupt"] = req.corrupt;
+	payload["numwant"] = req.num_want;
+
+	char str_key[9];
+	std::snprintf(str_key, sizeof(str_key), "%08X", req.key);
+	payload["key"] = str_key;
+
+	if (req.event != event_t::none)
+	{
+		static const char* event_string[] = { "completed", "started", "stopped", "paused" };
+		int event_index = static_cast<int>(req.event) - 1;
+		TORRENT_ASSERT(event_index >= 0 && event_index < 4);
+		payload["event"] = event_string[event_index];
+	}
+
+	payload["peer_id"] = latin1_utf8(req.pid);
+
+	json::array &offers_array = payload["offers"].emplace_array();
+	for (auto const& offer : req.offers)
+	{
+		json::object payload_offer;
+		payload_offer["offer_id"] = latin1_utf8(offer.id);
+		json::object &obj = payload_offer["offer"].emplace_object();
+		obj["type"] = "offer";
+		obj["sdp"] = offer.sdp;
+		offers_array.emplace_back(std::move(payload_offer));
+	}
+
+	json::string const data = json::to_string(payload);
```

Comment:
`data` will go out of scope, it's can't be passed to the `async_write()` call below

Reply:
Ow, good catch, I'm fixing it.